yarn-project/p2p/src/services/tx_collection/tx_collection.ts

```diff
@@ -0,0 +1,280 @@
+
+  /** Handler for the txs-added event from the tx pool */
+  protected readonly handleTxsAddedToPool: TxPoolEvents['txs-added'];
+
+  constructor(
+    private readonly reqResp: Pick<ReqRespInterface, 'sendBatchRequest'>,
+    private readonly nodes: TxSource[],
+    private readonly constants: L1RollupConstants,
+    private readonly txPool: TxPool,
+    private readonly config: TxCollectionConfig,
+    private readonly dateProvider: DateProvider = new DateProvider(),
+    telemetryClient: TelemetryClient = getTelemetryClient(),
+    private readonly log: Logger = createLogger('p2p:tx_collection_service'),
+  ) {
+    const collectionManager: TxCollectionManager = {
+      collect: (collectFn, requested, info) => this.collect(collectFn, requested, info),
+      getFastCollectionRequests: () => this.fastCollection.getFastCollectionRequests(),
+    };
+
+    this.instrumentation = new TxCollectionInstrumentation(telemetryClient, 'TxCollection');
+
+    this.slowCollection = new SlowTxCollection(
+      this.reqResp,
+      this.nodes,
+      collectionManager,
+      constants,
+      this.config,
+      this.dateProvider,
+      this.log,
+    );
+
+    this.fastCollection = new FastTxCollection(
+      this.reqResp,
+      this.nodes,
+      collectionManager,
+      this.config,
+      this.dateProvider,
+      this.log,
+    );
+
+    this.reconcileFoundTxsLoop = new RunningPromise(
+      () => this.reconcileFoundTxsWithPool(),
+      this.log,
+      this.config.txCollectionReconcileIntervalMs,
+    );
+
+    this.handleTxsAddedToPool = (args: Parameters<TxPoolEvents['txs-added']>[0]) => {
+      const { txs, source } = args;
+      if (source !== 'tx-collection') {
+        this.foundTxs(txs);
+      }
+    };
+
+    this.txPool.on('txs-added', this.handleTxsAddedToPool);
+  }
+
+  /** Starts all collection loops. */
+  public start(): Promise<void> {
+    this.slowCollection.start();
+    this.reconcileFoundTxsLoop.start();
+
+    // TODO(palla/txs): Collect mined unproven tx hashes for txs we dont have in the pool and populate missingTxs on startup
+    return Promise.resolve();
+  }
+
+  /** Stops all activity. */
+  public async stop() {
+    await Promise.all([this.slowCollection.stop(), this.fastCollection.stop(), this.reconcileFoundTxsLoop.stop()]);
+    if (this.handleTxsAddedToPool) {
+      this.txPool.removeListener('txs-added', this.handleTxsAddedToPool);
+    }
+  }
+
+  /** Force trigger the slow collection and reconciliation loops */
+  public async trigger() {
+    await Promise.all([this.reconcileFoundTxsLoop.trigger(), this.slowCollection.trigger()]);
+  }
+
+  /** Returns L1 rollup constants. */
+  public getConstants(): L1RollupConstants {
+    return this.constants;
+  }
+
+  /** Starts collecting the given tx hashes for the given L2Block in the slow loop */
+  public startCollecting(block: L2Block, txHashes: TxHash[]) {
+    return this.slowCollection.startCollecting(block, txHashes);
+  }
```

Comment:
Is it possible slow collection and fast collection overlaps for the same txs?

Reply:
They may for provers. A prover keeps all missing txs for mined unproven nodes in slow collection, but triggers fast collection once it needs to start proving.